### PR TITLE
Geo: Adds a set of no dependency geo classes for JDBC driver

### PR DESCRIPTION
**Pull request by imotov**
This set of geo classes is meant to be used in the JDBC driver to represent geo data and possibly as an intermediate format to pass geo shapes for indexing in #35320.

Relates to #35767 and #35320

### CODE REVIEW

**Review by alpar-t — COMMENTED** [review]
I'm missing some context here, but left some questions around the build.

File: libs/geo/build.gradle
Comment (atorok) [inline libs/geo/build.gradle]: This should be `:libs:geo-tests` which needs to be defined in  `settings.gradle` to keep eclipse happy. 

Note to self: We could simplify this setup for the latest Eclipse version (gradle/gradle#4802).

File: libs/geo/build.gradle
Comment (atorok) [inline libs/geo/build.gradle]: Is this needed ? Couldn't find any code that uses  it.

File: libs/geo/build.gradle
Comment (atorok) [inline libs/geo/build.gradle]: Are we planning to publish this jar e.x to maven central ?
Note that we'll have to update the release manager in that case.
We should only configure publishing if we do plan to publish it.

File: libs/geo/build.gradle
Comment (atorok) [inline libs/geo/build.gradle]: the project should be `:libs:geo-tests` and it needs to be created in `settings.gradle`

**Review by alpar-t — COMMENTED** [review]
Thanks @imotov this look much better. 
Left some more comments, sorry I didn't notice the publishing in the first round.

File: libs/geo/build.gradle
Comment (imotov) [reply libs/geo/build.gradle]: That's a good question. Its content will be shipped as a part of fat jdbc jar for SQL, but it will also used inside elasticsearch itself for geo operations. So that probably means that it needs to be published, correct?

Comment (atorok) [reply libs/geo/build.gradle]: That's right, don't forget  to add it to release manager.

File: settings.gradle
Comment (atorok) [inline settings.gradle]: There's another conditional for eclipse,  line 104 that needs to be updated.

File: libs/geo/build.gradle
Comment (atorok) [inline libs/geo/build.gradle]: NIT: you could have `project(':libs:geo').name = "elasticsearch-geo"` in `settings.gradle` instead of this. That would make the `publishing` block bellow unnecessary and it has the advantage of showing the module name in the IDE as well.

**Review by alpar-t — APPROVED** [review]
Thanks @imotov . One nit, the build LGMT otherwise.

**Comment by imotov** [comment]
@elasticmachine run gradle build tests 2

File: libs/geo/src/main/java/org/elasticsearch/geo/GeoUtils.java
Comment (iverase) [inline libs/geo/src/main/java/org/elasticsearch/geo/GeoUtils.java]: I have some concerns about this class. Should we make it package private? I don't think this methods should be used outside of this package.

File: libs/geo/src/main/java/org/elasticsearch/geo/geometry/GeometryVisitor.java
Comment (iverase) [inline libs/geo/src/main/java/org/elasticsearch/geo/geometry/GeometryVisitor.java]: I don't really understand why we have this class. There are no test for it and not sure if needed. Could you justify it?

File: libs/geo/src/main/java/org/elasticsearch/geo/GeoUtils.java
Comment (imotov) [reply libs/geo/src/main/java/org/elasticsearch/geo/GeoUtils.java]: Indeed. I will make it package private.

File: libs/geo/src/main/java/org/elasticsearch/geo/geometry/GeometryVisitor.java
Comment (imotov) [reply libs/geo/src/main/java/org/elasticsearch/geo/geometry/GeometryVisitor.java]: This is an implementation of the [Visitor](https://en.wikipedia.org/wiki/Visitor_pattern) pattern. The basic idea is to simplify adding new operations on Geometries, without constantly modifying and adding new functionality to the Geometry hierarchy and keeping it as lightweight as possible. 

As an example, I used it for [WKT serialization](https://github.com/imotov/elasticsearch/blob/9b784e24520d385ceac533673be28a9fc4c4650d/libs/geo/src/main/java/org/elasticsearch/geo/utils/WellKnownText.java#L71). Later I can add XContent serialization in another project. The alternative to this approach is to have a large 

```
if (obj instanceof This) { 
  doThis((This) obj); 
} elseif (obj instanceof That) {
  doThat((That) obj);
...
} else {
  throw new IllegalArgumentException("Unknown object " + obj);
}
```
The Visitor Pattern replaces this structure with Interface inheritance making it easier to identify all places that are using this structure, and making a shape a compile-time failure instead of runtime. I think it is a cleaner approach. This class is tested as part of toWKT code path, but you are right, I should add a designated test for it.

File: libs/geo/src/test/java/org/elasticsearch/geo/geometry/BaseGeometryTestCase.java
Comment (nknize) [inline libs/geo/src/test/java/org/elasticsearch/geo/geometry/BaseGeometryTestCase.java]: I think we should reuse lucene's `GeoTestUtil` for creating these random geometries. There was a lot of work put in to making sure the test suite creates adversarial geometries that have a tendency to expose weaknesses in the API. I don't think that should hold up the PR, though. So lets create a separate issue to migrate some of the random geometry generators over to lucene's geo test suite.

**Review by nknize — APPROVED** [review]
LGTM. I left a couple minor comments that I think could be added in a future PR so we don't hold this up any longer. The big thing would be to start moving our random geometry generators to lucene's `GeoTestUtil`. Lets open a separate issue for that.

File: libs/geo/src/main/java/org/elasticsearch/geo/GeoUtils.java
Comment (imotov) [reply libs/geo/src/main/java/org/elasticsearch/geo/GeoUtils.java]: Agree. I renamed it to GeometryUtils when I made changed its visibility.

File: libs/geo/src/test/java/org/elasticsearch/geo/geometry/BaseGeometryTestCase.java
Comment (imotov) [reply libs/geo/src/test/java/org/elasticsearch/geo/geometry/BaseGeometryTestCase.java]: I opened #37278 as a follow-up.

File: libs/geo/src/test/java/org/elasticsearch/geo/geometry/LinearRingTests.java
Comment (matriv) [inline libs/geo/src/test/java/org/elasticsearch/geo/geometry/LinearRingTests.java]: Why is it not extending the `BaseGeometryTestCase` ?

Comment (matriv) [reply libs/geo/src/test/java/org/elasticsearch/geo/geometry/LinearRingTests.java]: Ah Ignore... I see it's not supported by WKT...

File: libs/geo/src/main/java/org/elasticsearch/geo/geometry/Rectangle.java
Comment (matriv) [inline libs/geo/src/main/java/org/elasticsearch/geo/geometry/Rectangle.java]: shouldn't be a check for minLon maxLon as well?

File: libs/geo/src/main/java/org/elasticsearch/geo/geometry/GeometryVisitor.java
Comment (matriv) [inline libs/geo/src/main/java/org/elasticsearch/geo/geometry/GeometryVisitor.java]: you could remove `of`.

File: libs/geo/src/main/java/org/elasticsearch/geo/geometry/LinearRing.java
Comment (matriv) [inline libs/geo/src/main/java/org/elasticsearch/geo/geometry/LinearRing.java]: Minor: you could remove this empty line.

File: libs/geo/src/main/java/org/elasticsearch/geo/geometry/LinearRing.java
Comment (matriv) [inline libs/geo/src/main/java/org/elasticsearch/geo/geometry/LinearRing.java]: same here.

File: libs/geo/src/main/java/org/elasticsearch/geo/geometry/MultiLine.java
Comment (matriv) [inline libs/geo/src/main/java/org/elasticsearch/geo/geometry/MultiLine.java]: same here.

File: libs/geo/src/main/java/org/elasticsearch/geo/geometry/MultiPolygon.java
Comment (matriv) [inline libs/geo/src/main/java/org/elasticsearch/geo/geometry/MultiPolygon.java]: and here.

File: libs/geo/src/main/java/org/elasticsearch/geo/utils/WellKnownText.java
Comment (matriv) [inline libs/geo/src/main/java/org/elasticsearch/geo/utils/WellKnownText.java]: Why is `0` a whitespaceChar ?

File: libs/geo/src/main/java/org/elasticsearch/geo/geometry/Rectangle.java
Comment (iverase) [reply libs/geo/src/main/java/org/elasticsearch/geo/geometry/Rectangle.java]: Nope, it is perfectly valid to have minLon > maxLon, it just means that the Rectangle crosses the dateline.

File: libs/geo/src/main/java/org/elasticsearch/geo/utils/WellKnownText.java
Comment (imotov) [reply libs/geo/src/main/java/org/elasticsearch/geo/utils/WellKnownText.java]: Not only 0. Basically this line means that everything between 0 and ` ` will be treated as a whitespace character. This is to ignore all control characters below space ` ` including new line,  line feed, tabs, new page etc. by treating it as a whitespace character. I have decided to leave it as is instead of configuring each a subset of character manually since this is [the default](https://docs.oracle.com/javase/7/docs/api/java/io/StreamTokenizer.html#StreamTokenizer(java.io.InputStream)) behavior of StreamTokenizer, but we can be more restrictive here and allow only spaces, tabs and line feeds as suggested in the [standard](http://docs.opengeospatial.org/is/12-063r5/12-063r5.html#138) @iverase, @nknize what do you think?

Comment (iverase) [reply libs/geo/src/main/java/org/elasticsearch/geo/utils/WellKnownText.java]: The standard seems to only allow spaces, tabs or line feeds to improve readability so let's do as recommended.

Comment (imotov) [reply libs/geo/src/main/java/org/elasticsearch/geo/utils/WellKnownText.java]: :+1:

**Comment by imotov** [comment]
@matriv I pushed the changes that you have requested. Could you take another look?

**Review by matriv — APPROVED** [review]
LGTM. Thank you for your patience.

**Comment by karussell** [comment]
I'm curious why not use [JTS](https://github.com/locationtech/jts) instead? (BSD-like license, actively maintained, solid implementation, wide spread adoption, [without dependencies](https://github.com/locationtech/jts/tree/master/modules/core), ...)

**Pull request merged**
